packages/lesswrong/components/common/FlashMessages.tsx

```diff
@@ -1,105 +1,188 @@
 
 export const MessageContextProvider = ({children}: {
   children: ReactNode
 }) => {
-  const [messages,setMessages] = useState<AnyBecauseTodo[]>([]);
+  const [messages,setMessages] = useState<MessagesState>(() => ({
+    messages: [],
+    expiryTimer: null,
+    animationState: "closed",
+  }));
+  
+  const clear = useCallback(() => {
+    setMessages(messages => {
+      if (messages.expiryTimer) {
+        clearTimeout(messages.expiryTimer);
+      }
+      if (messages.animationState === "closing") {
+        return {
+          messages: [],
+          expiryTimer: null,
+          animationState: "closed",
+        };
+      } else {
+        return {
+          messages: messages.messages,
+          expiryTimer: setTimeout(clear, closeAnimationDelay),
+          animationState: "closing",
+        }
+      }
+    });
+  }, []);
   
   const flash = useCallback((message: WithMessagesMessage) => {
-    if (!messages.length) {
-      setMessages([message])
-    } else {
-      // Show the transition for clearing the old message, then pop up the new one
-      setMessages(messages.map((message: AnyBecauseTodo) => ({...message, hide:true})));
-      setTimeout(() => {
-        setMessages([message]);
-      }, 500);
-    }
-  }, [messages]);
+    setMessages(messages => {
+      if (messages.expiryTimer) {
+        clearTimeout(messages.expiryTimer);
+      }
+      return {
+        messages: [
+          ...messages.messages,
+          message
+        ],
+        expiryTimer: setTimeout(clear, messageExpirationTime),
+        animationState: messages.animationState === "closed" ? "opening" : "open",
+      };
+    });
+  }, [clear]);
 
-  const clear = useCallback(() => {
-    setMessages(messages.map((message: AnyBecauseTodo) => ({...message, hide:true})));
-    setTimeout(() => {
-      setMessages([]);
-    }, 500);
-  }, [messages]);
+  const tick = useCallback(() => {
+    setMessages(messages => {
+      if (messages.animationState === "opening") {
+        return {
+          ...messages,
+          animationState: "open",
+        };
+      } else {
+        return messages;
+      }
+    });
+  }, []);
 
   const messagesContext = useMemo(
-    () => ({ messages, flash, clear }),
-    [messages, flash, clear]
+    () => ({ flash, clear, tick }),
+    [flash, clear, tick]
   );
 
-  // FIXME: While this is mostly referentially stable, MessageContext will change
-  // when flash-messages are added or removed. This means that when a flash-message
-  // is added, every component on the page that called useMessages() will rerender,
-  // even if it was intended to be send-only. That includes a lot of components
-  // including vote buttons, so on a page with a lot of comments, this can be
-  // very slow.
-  
-  return <MessageContext.Provider value={messagesContext}>
-    {children}
-  </MessageContext.Provider>
+  return <MessageFunctionsContext.Provider value={messagesContext}>
```

Comment:
Not sure if this ever gets unmounted in a meaningful sense, but do we want to clear `expiryTimer` in state when it does?  i.e. 
```
useEffect(() => {
  return () => {
    if (state.expiryTimer !== null) clearTimeout(state.expiryTimeout);
  };
}, [state.expiryTimer]);
```

Reply:
I believe this component never unmounts (it's on `Layout`, which is reused across client-side navigations.)